### PR TITLE
Plugin editor

### PR DESCRIPTION
**Pull request by mariusandra**
## Changes

Edit plugins directly in the interface

image

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests

### CODE REVIEW

File: frontend/src/scenes/plugins/PluginCard.tsx [hunk 1]
```diff
@@ -113,7 +130,7 @@ export function PluginCard({ name, description, url, pluginConfig, pluginId, err
                             <Button
                                 type="primary"
                                 loading={loading}
-                                onClick={() => installPlugin(url, 'repository')}
+                                onClick={() => (url ? installPlugin(url, 'repository') : null)}
```
Comment (Twixes) [inline frontend/src/scenes/plugins/PluginCard.tsx:133]: We don't actually care about returning a value here, right? If so, the whole function should be ternary'd, i.e. ` onClick={url ? () => { installPlugin(url, 'repository') } : undefined}`

File: frontend/src/scenes/plugins/PluginDrawer.tsx [hunk 2]
```diff
@@ -47,114 +49,135 @@ export function PluginDrawer(): JSX.Element {
...
+                footer={
+                    <>
+                        <div style={{ display: 'flex' }}>
+                            <div style={{ flexGrow: 1 }}>
+                                {canDelete && (
+                                    <Popconfirm
+                                        placement="topLeft"
+                                        title="Are you sure you wish to uninstall this plugin?"
+                                        onConfirm={editingPlugin ? () => uninstallPlugin(editingPlugin.name) : () => {}}
+                                        okText="Yes"
```
Comment (Twixes) [inline frontend/src/scenes/plugins/PluginDrawer.tsx:68]: More "conscious" UX would be including the verb here, i.e. "Yes, uninstall" (conversely "No, cancel").

File: frontend/src/scenes/plugins/PluginCard.tsx [hunk 3]
```diff
@@ -49,6 +60,12 @@ export function PluginCard({ name, description, url, pluginConfig, pluginId, err
                 style={{ height: '100%', display: 'flex', marginBottom: 20 }}
                 bodyStyle={{ display: 'flex', flexDirection: 'column', flexGrow: 1 }}
             >
+                {pluginType === 'source' ? (
+                    <SourcePluginTag
+                        title="Source"
```
Comment (Twixes) [inline frontend/src/scenes/plugins/PluginCard.tsx:65]: "Source Code" as well perhaps.

**Review by Twixes — COMMENTED** [review]
Really cool! Just some notes for more polish.

File: frontend/src/scenes/plugins/install/SourcePlugin.tsx [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+export function SourcePlugin(): JSX.Element {
+    const { sourcePluginName, pluginError, loading } = useValues(pluginsLogic)
+    const { setSourcePluginName, installPlugin } = useActions(pluginsLogic)
+
+    return (
+        <div style={{ marginTop: 16 }}>
+            <Card>
+                <h3 className="l3">Write The Code</h3>
+                <p>Write your own plugin directly in PostHog! Just give it a name and start coding!</p>
```
[the remark this answers is not in the document]
Comment (mariusandra) [reply frontend/src/scenes/plugins/install/SourcePlugin.tsx]: Removed one sentence and added a link to the docs.

File: frontend/src/scenes/plugins/PluginSource.tsx [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                                            return Promise.resolve()
+                                        } catch (error) {
+                                            return Promise.reject('Not valid JSON!')
+                                        }
+                                    },
+                                },
+                            ]}
+                        >
+                            <MonacoEditor language="json" theme="vs-dark" height={200} />
+                        </Form.Item>
```
[the remark this answers is not in the document]
Comment (mariusandra) [reply frontend/src/scenes/plugins/PluginSource.tsx:104]: Made it a bit nicer.

File: frontend/src/scenes/plugins/PluginSource.tsx [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            }
+        >
+            <Form form={form} layout="vertical" onFinish={savePluginSource}>
+                {editingSource ? (
+                    <>
+                        <Form.Item label="Name" name="name" required rules={[requiredRule]}>
+                            <Input />
+                        </Form.Item>
+                        <Form.Item label="Source Code" name="source" required rules={[requiredRule]}>
+                            <MonacoEditor language="javascript" theme="vs-dark" height={400} />
```
[the remark this answers is not in the document]
Comment (mariusandra) [reply frontend/src/scenes/plugins/PluginSource.tsx]: There's a lot that I would like to do with this drawer including fixing the hardcoded heights, but I'd love to tackle it in a different PR. It's still beta ;).

**Comment by mariusandra** [comment]
Some things fixed with a comment, some without, and except for one issue (static heights) everything else should be fixed.

Since the interface might change dramatically between now and the next release, I'd propose to wait a bit with the height issue. Making this properly responsive is a lot of work... and I'd rather spend that time working on higher priority tickets :).

**Review by Twixes — APPROVED** [review]
👌

**Pull request merged**